### PR TITLE
(GRPO) Add importance sampling modes (TRL >= 0.26.0), Switch to vllm_importance_sampling_clip (support for TRL >= 1.6.0) and Fix VESPO not using the importance sampling ratio

### PR DESCRIPTION
**Pull request by marcandrelarochelle**
- Add importance sampling modes (TRL >= 0.26.0)
- Switch to vllm_importance_sampling_clip (support for TRL >= 1.6.0)
- Fix VESPO importance sampling ratio not using the the correct computed importance sampling

### CODE REVIEW

**Comment by danielhanchen** [comment]
Thanks, this is a good direction and the VESPO diagnosis is correct. I validated against real TRL sources (1.6.0, 1.7.1). The config wiring, defaults, `token_*` mask semantics, and the switch to `getattr(trainer.args, "vllm_importance_sampling_cap", None)` (which also fixes a latent AttributeError, since 1.6.0+ store cap on `args`, not the trainer) all look right. Three issues to fix.

## 1. Sequence modes compute a sum of exp ratios instead of a product (affects the default mode)

TRL aggregates in log space, then exponentiates once. TRL 1.6.0 `grpo_trainer.py:2075-2084` (identical in 1.7.1):

```python
per_token_logps_diff = (old_per_token_logps - sampling_per_token_logps) * mask
sequence_level_is = self.vllm_importance_sampling_mode in ["sequence_mask", "sequence_truncate"]
if sequence_level_is:
    per_sequence_logps_diff = per_token_logps_diff.sum(dim=-1, keepdim=True)
    logps_diff = per_sequence_logps_diff
else:
    logps_diff = per_token_logps_diff
vllm_importance_sampling_ratio = torch.exp(logps_diff)
```

So TRL's sequence ratio is `exp(sum_t (old - sampling))` = the product of per-token ratios (a single value near 1). The PR does:

```python
importance_sampling_ratio = torch.exp((old * mask) - sampling_per_token_logps)   # exp first
if vllm_importance_sampling_mode in ["sequence_mask", "sequence_truncate"]:
    importance_sampling_ratio = importance_sampling_ratio.sum(dim=-1, keepdim=True)  # sum of exp
```

`sum(exp(x)) != exp(sum(x))`. For a T-token sequence with per-token ratios near 1, TRL gives about 1 while this gives about T, so with the default `sequence_mask` + `clip_max=3.0` most sequences longer than ~3 tokens exceed the cap and get zeroed by `masked_fill`. Since `sequence_mask` is the default mode, the standard config is affected.

There is a second, related detail: TRL masks the difference `(old - sampling) * mask`, so prompt/pad positions contribute 0 to the sum. Here the mask is applied only to `old` (`(old * mask) - sampling`), so masked positions add `exp(-sampling)` into the per-sequence sum. For token modes the later `* mask` cancels it, but the sequence sum bakes it in.

Suggested fix (mirror TRL, log space):

```python
logps_diff = (old - sampling_per_token_logps) * mask
if vllm_importance_sampling_mode in ["sequence_mask", "sequence_truncate"]:
    logps_diff = logps_diff.sum(dim=-1, keepdim=True)
importance_sampling_ratio = torch.exp(logps_diff)
# then apply the truncate/mask clipping as you already do
```

## 2. VESPO now applies the importance sampling ratio twice

Passing the local ratio into `get_gamma_weights` is the right fix (on main it was always `None`). But `get_gamma_weights` applies the correction internally, so TRL excludes VESPO from the post-loss multiply. TRL 1.6.0 `grpo_trainer.py:2612-2613` (same in 1.7.1):

```python
if self.use_vllm and self.vllm_importance_sampling_correction and self.loss_type != "vespo":
    per_token_loss = per_token_loss * inputs["importance_sampling_ratio"]
```

The unsloth multiply has no vespo guard:

```python
if use_vllm and sampling_per_token_logps is not None:
    loss_i = loss_i * importance_sampling_ratio
```

So after this PR, VESPO gets the ratio once inside `get_gamma_weights` and again here. Add `and loss_type != "vespo"` to that condition to match TRL.

## 3. Minor: `mode=None` raises for a pre-0.26.0 TRL band

`grpo_accumulated_loss` sets `kwargs["vllm_importance_sampling_mode"] = getattr(trainer.args, "vllm_importance_sampling_mode", None)` unconditionally. When that arg does not exist, the key is present with value `None`, so `kwargs.get("vllm_importance_sampling_mode", "sequence_mask")` in `grpo_compute_loss` returns `None` (the default is dead), and with IS active it falls through to `raise ValueError("Unknown vLLM importance sampling mode: None")`. The `cap -> clip` shim does not set a mode. 0.27.2/0.28.0 do have the field so only the older band is affected; defaulting `mode` to `"token_truncate"` when it resolves to `None` (legacy `clamp(max=cap)` semantics) would cover it.

Items 1 and 2 are the ones that change training behavior.

**Comment by danielhanchen** [comment]
Pushed the fixes directly (630bd22):

1. Sequence modes now aggregate log-ratios then exp (product), matching TRL, instead of summing exp ratios.
2. Skip the post-loss IS multiply for `vespo` (it already applies the ratio inside `get_gamma_weights`), avoiding a double count.
3. Default the mode to `token_truncate` when older TRL lacks the arg, so the legacy `cap` path does not hit the unknown-mode error.

Nice catch on the VESPO ratio. Thanks.

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -428,13 +431,39 @@ def grpo_compute_loss(
...
+                importance_sampling_ratio > max_val
+        )
+
+        importance_sampling_ratio = importance_sampling_ratio.masked_fill(
+                invalid_mis_mask, value=0.0
+        )
+    else:
+        raise ValueError(
+                f"Unknown vLLM importance sampling level: {vllm_importance_sampling_mode}. Possible values are 'token_truncate', 'token_mask', 'sequence_truncate', and 'sequence_mask'."
+        )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: On the current head importance_sampling_ratio is initialized to None above the block and computed inside the use_vllm and sampling_per_token_logps is not None guard, so there is no NameError and the clipping is not unconditional.

File: unsloth_zoo/rl_replacements.py [hunk 2]
```diff
@@ -751,7 +780,10 @@ def grpo_accumulated_loss(
 
     # Pop from kwargs to avoid downstream issues.
     _ = kwargs.pop("sampling_per_token_logps", None)
-    kwargs["vllm_importance_sampling_cap"] = trainer.vllm_importance_sampling_cap if sampling_per_token_logps is not None else None
+    kwargs["vllm_importance_sampling_cap"] = trainer.args.vllm_importance_sampling_cap if hasattr(trainer.args, "vllm_importance_sampling_cap") is not None else None
+    kwargs["vllm_importance_sampling_mode"] = trainer.args.vllm_importance_sampling_mode if hasattr(trainer.args, "vllm_importance_sampling_mode") is not None else None
+    kwargs["vllm_importance_sampling_clip_min"] = trainer.args.vllm_importance_sampling_clip_min if hasattr(trainer.args, "vllm_importance_sampling_clip_min") is not None else None
+    kwargs["vllm_importance_sampling_clip_max"] = trainer.args.vllm_importance_sampling_clip_max if hasattr(trainer.args, "vllm_importance_sampling_clip_max") is not None else None
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: The code already uses getattr(trainer.args, ..., None); there is no hasattr(...) is not None pattern here, so this does not apply to the current head.

**Comment by danielhanchen** [comment]
@codex review

**Review by danielhanchen — APPROVED** [review]
Independently re-reviewed against TRL 1.5.0/1.6.0/1.7.1/main: the log-space importance-sampling ratio is a byte-for-byte TRL port (allclose across all four modes), the VESPO double-apply fix matches TRL's own loss_type != vespo gating with the ratio routed into get_gamma_weights, version handling is via getattr fallbacks (no off-by-one gate), and non-IS runs are unchanged. Verified the combined result with #869 auto-merges cleanly and stays valid. Clean.
